docs/source/notes/randomness.rst

```diff
@@ -2,70 +2,73 @@ Reproducibility
+    torch.set_deterministic(True)
 
-A number of operations have backwards kernels that use :attr:`atomicAdd`,
-including :meth:`torch.nn.functional.embedding_bag`,
-:meth:`torch.nn.functional.ctc_loss`, :meth:`torch.nn.functional.interpolate`,
-and many forms of pooling, padding, and sampling.
+CUDA convolution
+----------------
+When running convolution operations CUDA devices, two options must be set::
 
-There is currently no simple way of avoiding nondeterminism in these functions.
-
-Additionally, the backward path for :meth:`repeat_interleave` operates
-nondeterministically on the CUDA backend because :meth:`repeat_interleave`
-is implemented using :meth:`index_select`, the backward path for
-which is implemented using :meth:`index_add_`, which is known to operate
-nondeterministically (in the forward direction) on the CUDA backend (see above).
+    torch.backends.cudnn.deterministic = True
+    torch.backends.cudnn.benchmark = False
 
-cuDNN
-.....
-When running on the cuDNN backend, two further options must be set::
+Alternatively, you can use the following, since CUDA convolution will be
+deterministic if either :code:`torch.set_deterministic(True)` or
+:code:`torch.backends.cudnn.deterministic = True` is set::
 
-    torch.backends.cudnn.deterministic = True
+    torch.set_deterministic(True)
     torch.backends.cudnn.benchmark = False
 
 On some versions of cuDNN and CUDA, RNN and LSTM may have non-deterministic behavior.
 See :meth:`torch.nn.RNN` and :meth:`torch.nn.LSTM` for details and workarounds.
 
+When :code:`torch.backends.cudnn.benchmark = True` is set and CUDA's convolution
+operation is called with a new set of parameters, multiple different implementations
+will be run and benchmarked. The fastest implementation will then be cached and used
+for all further CUDA convolution operations with that same set of parameters until
+the PyTorch program quits. So if you need each execution of your script to produce
+the same CUDA convolution results, benchmarking must be turned off. However, if you
+only care about determinism within one execution of the script, benchmarking
+can be turned on for better performance.
+
 .. warning::
 
-    Deterministic operation may have a negative single-run performance impact,
-    depending on the composition of your model. Due to different underlying
-    operations, which may be slower, the processing speed (e.g. the number of
-    batches trained per second) may be lower than when the model functions
-    nondeterministically. However, even though single-run speed may be
-    slower, depending on your application determinism may save time by
-    facilitating experimentation, debugging, and regression testing.
+    Deterministic operations are often slower than nondeterministic operations, so
+    single-run performance may decrease for your model. However, determinism may
+    save time in development by facilitating experimentation, debugging, and
+    regression testing.
 
 Numpy
```

Comment:
I suggest folding this section into the section on setting random numbers. Is this section up-to-date with modern NumPy random behavior, which suggests acquiring a generator object before performing some operations? Maybe the sentence can just be made broader to discuss that inherent sources of randomness, like NumPy's generators, should be initialized consistently? (That is, not just setting the seed but using the same generation algorithm.)

Reply:
I added a note that if the user is using NumPy's generator objects, they should set those seeds. Do you think this is good?